src/ripple/app/tx/impl/InvariantCheck.h

```diff
@@ -365,6 +365,32 @@ class NFTokenCountTracking
         beast::Journal const&);
 };
 
+/**
+ * @brief Invariant: Trust line balance cannot be negative after Clawback.
+ *
+ * We iterate all the trust lines affected by this transaction and ensure
+ * that they have non-negative balance.
+ */
+class ValidClawback
+{
+    std::uint32_t trustlinesChanged = 0;
+
+public:
+    void
+    visitEntry(
+        bool,
+        std::shared_ptr<SLE const> const&,
+        std::shared_ptr<SLE const> const&);
+
+    bool
+    finalize(
```

Comment:
On a related note, can we force all the `InvariantChecks` to adhere to a common interface by inheriting from an abstract class? It appears that all of them have `visitEntry` and `finalize` methods.

Reply:
No, I don't know this.  I was not in on the design of the invariants.  But inheriting from an abstract class doesn't really _enforce_ a common interface.  The derived class can add whatever methods it wants.  It also is not _required_ to implement inherited methods well.

The thing that enforces the interface of these classes is that they are all consumed by `ApplyContext::checkInvariantHelper()` which expects a regular interface: https://github.com/XRPLF/rippled/blob/develop/src/ripple/app/tx/impl/ApplyContext.cpp#L91-L97

If the invariant does not comply with the interface expected by `checkInvariantsHelper()` then `checkInvariantsHelper()` will fail to compile.

This kind of design is common in template-heavy code, as opposed to code that is inheritance heavy.  Templates (largely) rely on "duck typing".  If it walks like a duck and quacks like a duck, then it is a duck.  All of the InvariantCheckers walk like InvariantCheckers and quack like InvariantCheckers so, as far as `checkInvariantsHelper()` is concerned, they are InvariantCheckers.

I hope that helps.

Reply:
thanks for the explanation!